### PR TITLE
[inductor] Add a lowering for aten.sym_storage_offset

### LINKED ISSUE #198068: AOTInductor drops the storage offset of a mutable view argument under dynamic shapes (missing lowering for aten.sym_storage_offset)
### 🐛 Describe the bug

When a custom operator declares a mutable argument (`mutates_args=...`, i.e. a `Tensor(a!)` in the schema) and that argument is a **view of a larger buffer**, AOTInductor drops the view's storage offset under **dynamic shapes**. Every write lands at the start of the base buffer instead of at the view's offset.

There is no error or warning. The compiled model just returns wrong numbers, so it looks like a numerical problem in the model rather than a compiler issue.

Static export is unaffected, and so is `torch.compile` with the default Python wrapper. Only dynamic-shape AOTInductor is wrong.

#### Repro

Reproduces on **CPU and CUDA**, so no GPU is needed.

```python
"""AOTInductor drops the storage offset of a mutable (Tensor(a!)) view argument
under dynamic shapes.

    python repro.py         # cuda
    python repro.py cpu     # cpu
"""
import sys
import torch
from torch import nn
from torch.export import Dim
from torch._inductor import aoti_compile_and_package
from torch._inductor.package import load_package

DEV = sys.argv[1] if len(sys.argv) > 1 else "cuda"


@torch.library.custom_op("repro::write", mutates_args=("out",))
def write(out: torch.Tensor, x: torch.Tensor) -> None:
    out.copy_(x)


@write.register_fake
def _(out, x):
    return None


class M(nn.Module):
    """Two mutable writes into two disjoint slices of one zeroed buffer."""

    def forward(self, x0, x1):
        n0, n1 = x0.shape[0], x1.shape[0]
        buf = torch.zeros(n0 * 8 + n1 * 8, dtype=x0.dtype, device=x0.device)
        v0 = buf[: n0 * 8].view(n0, 8)      # storage_offset 0
        v1 = buf[n0 * 8 :].view(n1, 8)      # storage_offset n0*8, symbolic
        torch.ops.repro.write(v0, x0)
        torch.ops.repro.write(v1, x1)
        return v0.sum(), v1.sum()


def main():
    m = M().eval()
    xs = (torch.ones(10, 8, device=DEV), torch.full((5, 8), 2.0, device=DEV))
    expected = tuple(t.item() for t in m(*xs))          # (80.0, 80.0)

    for label, dynamic_shapes in (
        ("static ", None),
        ("dynamic", ({0: Dim("n0", min=2, max=64)}, {0: Dim("n1", min=2, max=64)})),
    ):
        ep = torch.export.export(m, xs, dynamic_shapes=dynamic_shapes)
        pkg = aoti_compile_and_package(ep, package_path=f"/tmp/repro_{label.strip()}.pt2")
        got = tuple(t.item() for t in load_package(pkg)(*xs))
        print(f"{label}  expected={expected}  aoti={got}  {'OK' if got == expected else 'WRONG'}")


if __name__ == "__main__":
    print(f"torch {torch.__version__}  device {DEV}")
    main()
```

#### Output

```
torch 2.14.0+cu130  device cuda
static   expected=(80.0, 80.0)  aoti=(80.0, 80.0)  OK
dynamic  expected=(80.0, 80.0)  aoti=(120.0, 0.0)  WRONG

torch 2.14.0+cu130  device cpu
static   expected=(80.0, 80.0)  aoti=(80.0, 80.0)  OK
dynamic  expected=(80.0, 80.0)  aoti=(120.0, 0.0)  WRONG
```

Both writes went to the first slice: it sums to 120 instead of 80, and the second slice was never written, so it sums to 0.

#### Root cause

`aten.sym_storage_offset` has no Inductor lowering.

1. Functionalization turns the mutating call into `auto_functionalized_v2`, recording the mutated view as a base plus its size, stride and storage offset.
2. The re-inplacing pass rebuilds the view as `as_strided(base, size, stride, sym_storage_offset(view))`. The post-grad FX graph is correct at this point, e.g.

   ```
   sym_storage_offset_default: "Sym(16384*(((s4 + 255)//256)))" = torch.ops.aten.sym_storage_offset.default(slice_2)
   as_strided_default = torch.ops.aten.as_strided.default(full, [...], [64, 1], sym_storage_offset_default)
   ```
3. `aten.sym_storage_offset.default` is not in `torch/_inductor/lowering.py`, while its siblings `sym_size.int`, `sym_stride.int` and `sym_numel` all are. Those exist precisely so the value is read from the node's metadata instead of being recomputed.
4. So it takes the implicit fallback, which re-executes the op on tensors rebuilt by `ir_node_to_tensor()`. That helper allocates with `torch.empty_strided(size, stride, ...)`, copying size and stride from the layout but never the offset, so every rebuilt tensor reports `storage_offset() == 0`.
5. The node lowers to the integer `0`, `as_strided` builds a `ReinterpretView` at offset 0, and the generated C++ passes `0L`:

   ```cpp
   // mutable argument, dynamic build
   reinterpret_tensor_wrapper(buf0, 2, int_array_2, int_array_1, 0L)
   // read-only argument in the same call, correct
   reinterpret_tensor_wrapper(buf0, 2, int_array_2, int_array_1,
                              16384L*c10::div_floor_integer(255L + s33, 256L))
   ```

Static export escapes this because the offset is a constant that never becomes a node. The default Python wrapper escapes it because the mutation stays in place and no `as_strided` is emitted.

Instrumenting the lowering shows the node's metadata is right and the lowered value is not:

```
[run_node] sym_storage_offset_default: meta_val=16384*(((s4 + 255)//256))  ->  lowered=0 (int)
```

#### Suggested fix

Register the missing lowering next to its siblings in `torch/_inductor/lowering.py`:

```python
@register_lowering(aten.sym_storage_offset.default)
def sym_storage_offset(a):
    val = V.graph.current_node.meta["val"]
    if isinstance(val, torch.SymInt):
        return val.node.expr
    else:
        return int(val)
```

With that patch applied, the repro above prints `OK` for both static and dynamic on CPU and CUDA. It also fixes a real model where several accumulator buffers are packed into one allocation and each consumer is handed its own slice.

A PR with this fix and a regression test in `test/inductor/test_aot_inductor.py` follows.

### Versions

Reproduced on 2.11.0+cu130, 2.13.0+cu130 and 2.14.0+cu130, on both CPU and CUDA (B200, CUDA 13.0, Python 3.12 and 3.14).

`torch/_inductor/lowering.py` on `main` registers `sym_size.int`, `sym_stride.int` and `sym_numel`, and contains no occurrence of `sym_storage_offset`, so `main` looks affected as well.

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @bobrenjc93 @aditvenk @laithsakka @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @avikchaudhuri @zhxchen17 @tugsbayasgalan @angelayi @ydwu4 @desertfire @yushangdi @iupaikov-amd @chenyang78 @eellison

### PR DESCRIPTION
**Pull request by jkzhang7**
Fixes #198068

`aten.sym_storage_offset` has no Inductor lowering, while its siblings `sym_size.int`, `sym_stride.int` and `sym_numel` all have one. Those exist precisely so the value is read from the node's metadata rather than recomputed.

Without a lowering the op takes the implicit fallback and is re-executed on a tensor rebuilt by `ir_node_to_tensor()`. That helper allocates with `torch.empty_strided(size, stride, ...)`, copying size and stride off the layout but never the offset, so every rebuilt tensor reports `storage_offset() == 0`.

That matters because the re-inplacing pass rebuilds a mutated view argument as `as_strided(base, size, stride, sym_storage_offset(view))`. Under dynamic shapes the offset is a symbolic node, it lowers to the integer `0`, and every write to a view of a larger buffer lands at the start of that buffer. The post-grad FX graph is correct; only the lowered value is wrong:

```
[run_node] sym_storage_offset_default: meta_val=16384*(((s4 + 255)//256))  ->  lowered=0 (int)
```

and the generated C++ shows it clearly, with the read-only argument in the same call keeping its offset:

```cpp
// mutable argument
reinterpret_tensor_wrapper(buf0, 2, int_array_2, int_array_1, 0L)
// read-only argument, correct
reinterpret_tensor_wrapper(buf0, 2, int_array_2, int_array_1,
                           16384L*c10::div_floor_integer(255L + s33, 256L))
```

Static export is unaffected because the offset is then a constant that never becomes a node, and the default Python wrapper is unaffected because the mutation stays in place and no `as_strided` is emitted. So this only shows up in dynamic-shape AOTInductor, and it shows up silently: no error, just wrong numbers.

## The fix

Read the value off the node's metadata, matching `sym_size` and `sym_stride`:

```python
@register_lowering(aten.sym_storage_offset.default)
def sym_storage_offset(a):
    val = V.graph.current_node.meta["val"]
    if isinstance(val, torch.SymInt):
        return val.node.expr
    else:
        return int(val)
```

## Test

`test_dynamic_mutated_view_storage_offset` in `test/inductor/test_aot_inductor.py` has a custom op mutate two disjoint slices of one zeroed buffer, with both row counts dynamic, so the second slice's storage offset is symbolic. It runs through the existing `check_model` path, so it covers the device variants the suite already parametrizes, CPU included.

Before the fix the two slices come back as `120.0` and `0.0` instead of `80.0` and `80.0`: both writes went to the first slice and the second was never written. After the fix they match eager.

## Validation

- The reproducer in #198068 prints `OK` for both static and dynamic, on CPU and on CUDA, with the patch applied.
- The new test's model was checked directly against an installed 2.14 build: it fails without the lowering and passes with it, on CPU and CUDA.
- Verified on 2.11.0+cu130, 2.13.0+cu130 and 2.14.0+cu130; all three are affected, so this is not a recent regression.
- Also fixes a real model where several accumulator buffers are packed into one allocation and each consumer is handed its own slice. Its compiled output went from a mean absolute error of 0.038 against eager to 0.0027, against a run-to-run noise floor of 0.0026.

## Notes

- I do not have a from-source build here, so the change was validated by applying it to installed wheels rather than through a local `lintrunner` and test run. The two added blocks are clean under the repo's `ruff format` and `ruff check` configuration; CI coverage would be welcome.
- The new test is not listed in `CPU_TEST_FAILURES`, `GPU_TEST_FAILURES` or `MPS_TEST_FAILURES`, so it is expected to pass everywhere. I could only exercise CPU and CUDA locally. If MPS turns out not to support a proxy-executor fallback with a symbolic storage offset, in the way `test_fallback_kernel_with_symexpr_output` already does not, I am happy to add an `MPS_TEST_FAILURES` entry.
- Happy to add a `release notes:` label if a maintainer tells me which one fits; `release notes: inductor (aoti)` looks right.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -9025,6 +9025,19 @@ def sym_numel(a):
...
 
 
+@register_lowering(aten.sym_storage_offset.default)
+def sym_storage_offset(a):
+    # Read the value off the node's metadata, like sym_size / sym_stride above.
+    # Without a lowering this op falls back to being re-executed on a tensor
+    # rebuilt by ir_node_to_tensor(), which allocates with torch.empty_strided()
+    # and therefore always reports a storage offset of 0.
+    val = V.graph.current_node.meta["val"]
+    if isinstance(val, torch.SymInt):
```
Comment (eellison) [inline torch/_inductor/lowering.py]: this should be using `a` the Tensor object offset . not the meta.

Comment (jkzhang7) [reply torch/_inductor/lowering.py]: Thanks for the review! Switched to reading the offset from the tensor instead of the meta.
